File: api/src/Domain/ProductCatalog/Brands/Services/BrandService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using DetailingArsenal.Domain.Users;

namespace DetailingArsenal.Domain.ProductCatalog {
    public interface IBrandService : IService {
        Task<Brand> GetById(Guid id);
        Task<Brand> Create(BrandCreate create, User user);
        Task<Brand> Update(Brand brand, BrandUpdate update, User user);
        Task Delete(Brand brand, User user);
    }

    public class BrandService : IBrandService {
        IBrandRepo repo;
        BrandNameUniqueSpecification unique;

        public BrandService(IBrandRepo repo, BrandNameUniqueSpecification unique) {
            this.repo = repo;
            this.unique = unique;
        }

        public async Task<Brand> GetById(Guid id) {
            var brand = await repo.FindById(id);
            return brand ?? throw new EntityNotFoundException();
        }

        public async Task<Brand> Create(BrandCreate create, User user) {
            var brand = new Brand(create.Name);

            await unique.CheckAndThrow(brand);
            await repo.Add(brand);

            return brand;
        }

        public async Task<Brand> Update(Brand brand, BrandUpdate update, User user) {
            brand.Name = update.Name;

            await unique.CheckAndThrow(brand);
            await repo.Update(brand);
            return brand;
        }

        public async Task Delete(Brand brand, User user) {
            await repo.Delete(brand);
        }
    }
}

File: app/src/api/product-catalog/brands/services/brand-service.ts
import { Brand } from '@/api';
import { http } from '@/api/core/http';
import { BrandCreate } from '../data-transfer-objects/brand-create';
import { BrandUpdate } from '../data-transfer-objects/brand-update';

export class BrandService {
    async get() {
        try {
            const res = await http.get('product-catalog/brand');

            if (res.data == null) {
                return [];
            }

            return res.data.map((s: any) => this._map(s));
        } catch (e) {
            if (e.response == null || e.response.status == 404) {
                return [];
            } else {
                throw e;
            }
        }
    }

    async create(create: BrandCreate) {
        const res = await http.post('product-catalog/brand', create);
        const b = this._map(res.data);
        return b;
    }

    async update(update: BrandUpdate) {
        const res = await http.put(`product-catalog/brand/${update.id}`, update);
        const b = this._map(res.data);
        return b;
    }

    async delete(id: string) {
        await http.delete(`product-catalog/brand/${id}`);
    }

    _map(b: any): Brand {
        return new Brand(b.id, b.name);
    }
}
